common/app/uiComponent/UIComponent.scala

```diff
@@ -0,0 +1,25 @@
+package uiComponent
+
+import model.ApplicationContext
+import play.api.libs.json.JsValue
+import play.api.mvc.Result
+import play.api.mvc.Results._
+import uiComponent.core.JavascriptRenderer
+
+import scala.util.{Failure, Success}
+
+trait UIComponentState {
+  def asJson: JsValue
+}
+
+abstract class UIComponent[-S <: UIComponentState](javascriptFile: String) {
```

Comment:
@TBonnin Might decoupling using typeclasses be better than using inheritance with type constraints? It seems to fit more naturally here; it allows completely independent types to represent different components as they wouldn't have to directly inherit anything.